src/ext.rs

```diff
@@ -0,0 +1,145 @@
+use core::{marker::PhantomData, task::Poll};
+
+use crate::{
+    api::{reply, request},
+    backend::{Backend, CoreOnly, Dispatch, NoId},
+    client::{ClientError, ClientImplementation, FutureResult, PollClient},
+    error::Error,
+    platform::{Platform, Syscall},
+    postcard_deserialize, postcard_serialize_bytes,
+    service::ServiceResources,
+    types::{Context, CoreContext},
+};
+
+use serde::{de::DeserializeOwned, Serialize};
+
+pub trait Extension {
+    type Request: DeserializeOwned + Serialize;
+    type Reply: DeserializeOwned + Serialize;
+}
+
+/// Dispatches extension requests to custom backends.
+pub trait ExtensionDispatch<P: Platform>: Dispatch<P> {
```

Comment:
(The following is based on the feeling that cfg-if'ing `DispatchRequirement` leads to too much of a "parallel worlds" split of upstream Trussed)

I was thinking this would look more like:

```rust
pub trait Dispatch<P> {
    /// The ID type for the custom backends used by this dispatch implementation.
    type BackendId: 'static;
    /// The context type used by this dispatch.
    type Context: Default;
    /// The ID type for the extensions supported by this dispatch implementation.
    #[cfg(feature = "api-extensions")]
    type ExtensionId: TryFrom<u8, Error = Error>;

    fn request(
        &mut self,
        backend: &Self::BackendId,
        ctx: &mut Context<Self::Context>,
        request: &Request,
        resources: &mut ServiceResources<P>,
    ) -> Result<Reply, Error> { ... }

    #[cfg(feature = "api-extensions")]
    fn extension_request(
        &mut self,
        backend: &Self::BackendId,
        extension: &Self::ExtensionId,
        ctx: &mut Context<Self::Context>,
        request: &request::Extension,
        resources: &mut ServiceResources<P>,
    ) -> Result<reply::Extension, Error> { ... }
}
```

Might not be so nice because 2532-associated-type-defaults is unstable (for ExtensionId).
On the other hand, Trussed doesn't really semantically understand ExtensionId anyway, so it could just pass it along untouched:

```rust
pub trait Dispatch<P> {
    /// The ID type for the custom backends used by this dispatch implementation.
    type BackendId: 'static;
    /// The context type used by this dispatch.
    type Context: Default;

    fn request(
        &mut self,
        backend: &Self::BackendId,
        ctx: &mut Context<Self::Context>,
        request: &Request,
        resources: &mut ServiceResources<P>,
    ) -> Result<Reply, Error> { ... }

    #[cfg(feature = "api-extensions")]
    fn extension_request(
        &mut self,
        backend: &Self::BackendId,
        ctx: &mut Context<Self::Context>,
        // extension: u8,
        request: &request::Extension,
        resources: &mut ServiceResources<P>,
    ) -> Result<reply::Extension, Error> { 
        let _ = (backend, /*extension,*/ ctx, request, resources);
        Err(Error::RequestNotAvailable)
    }
}
```

I commented out the `extension: u8` as the `request::Extension` already contains it. But now at this point, there isn't an actual difference between `fn request` and `fn extension_request`, so... The entire `api-extensions` feature is seems moot and we can just stick with a single `request` method?

I think you're trying to add some type safety via the `ExtensionId`, but since in `request::Extension` you're actually passing the naked `u8` with no indication of its type, it seems this gives no additional safety anyway?

In which case (while staying with this simple approach, and assuming you have desire for multiple different ExtensionId types), perhaps that API would be better as 
```
        Extension:
          - id: Bytes<{ some small number }>,  // <- serializing more metadata
          - request: Bytes<{ crate::config::EXTENSION_REQUEST_LENGTH }>
```
or even
```
Extension:
  kind: u8,  // <- indicating the underlying type implementing ExtensionId
  id: u8,
  request: Bytes
```

Reply:
The above is for the backend + service side of things, I do think it would be nice to have upstream code factoring out ergonomic and type safe-er client struct APIs.